### PR TITLE
Latest sls support + various updates

### PR DESCRIPTION
**Pull request by Enase**
- provider verification added
- schema validation for plugin custom properties added
- schema validation for plugin function properties added
- tslint replaced by eslint, code fixed to meet code style requirements
- tests fixed according to current latest serverless versions
- editor config added
- dependencies updated

Fixes:
- https://github.com/functionalone/serverless-iam-roles-per-function/issues/51

### CODE REVIEW

**Comment by Enase** [comment]
@glicht @theburningmonk @randomhash @andersquist

Feel free to add comments and discuss..

**Comment by andersquist** [comment]
Great job! 👍

Just a few things that comes to my mind when looking thru the code. If we're enforcing formatting with `prettier`/`editorconfig` or similar tooling I think it would be bit easier if that is the only thing that is done in a PR because otherwise it could be hard to perform a code review since it will make a lot of changes/diffs that are not actual functionality changes.

I think we're using Conventional commits on this repository - that is prefixing a commit with either: `doc:`,`feat:`, `chore:` etc. @glitcht could perhaps confirm this. In any case you can find the full specification [here](https://www.conventionalcommits.org/en/v1.0.0/) if you want to read about it. The benefit is that we could setup `sematic-release` to handle versioning and updating a changelog etc further on.

**Comment by Enase** [comment]
@andersquist Thanks for help with `sematic-release`, I've changed commit titles and squashed some of them.

**Comment by Enase** [comment]
@glicht @theburningmonk @randomhash @andersquist
Could I leave it as is or any actions required?

File: src/lib/index.ts [hunk 1]
```diff
@@ -23,8 +22,43 @@ class ServerlessIamPerFunctionPlugin {
    * @param options
    */
   constructor(serverless: any) {
-    this.provider = 'aws';
     this.serverless = serverless;
+
+    if (this.serverless.service.provider.name !== 'aws') {
+      throw new this.serverless.classes.Error('ses-template plugin supports only AWS');
```
Comment (CodeVision) [inline src/lib/index.ts:28]: Might want to fix the plugin name here :smile:

Comment (Enase) [reply src/lib/index.ts]: sorry, my bad. fixed

Comment (Enase) [reply src/lib/index.ts]: thanks for review. let me know if you see any other issues

**Comment by Enase** [comment]
@glicht @theburningmonk @randomhash @andersquist

Any updates?

**Comment by randomhash** [comment]
@Enase sorry, it's a big pack of changes, I will get to this pr in few days

File: .travis.yml [hunk 2]
```diff
@@ -18,6 +19,10 @@ before_install:
 
 install:
   - travis_retry npm install
-  - npm install --no-save --ignore-scripts `npx npm-get-version serverless@$SERVERLESS_VERSION`
+  - travis_retry npm install --no-save --ignore-scripts `npx npm-get-version serverless@$SERVERLESS_VERSION`
+
+script:
+  - npm run lint
```
Comment (randomhash) [inline .travis.yml:25]: if you adding linter, we should fail CI on not correctly linted code at least

File: package.json [hunk 3]
```diff
@@ -36,26 +37,29 @@
...
-    "@serverless/enterprise-plugin": "^1.3.10",
-    "@types/chai": "^4.2.0",
-    "@types/lodash": "^4.14.138",
-    "@types/mocha": "^5.2.7",
-    "@types/node": "^6.14.7",
+    "@serverless/enterprise-plugin": "^4.1.1",
+    "@types/chai": "^4.2.14",
+    "@types/lodash": "^4.14.164",
+    "@types/mocha": "^8.0.3",
+    "@types/node": "^14.14.6",
```
Comment (randomhash) [inline package.json:47]: looks like we should leave it at 12 for now

**Comment by randomhash** [comment]
what do you think about locking dependencies on the current version for non-critical stuff? so it not get broken if something will be corrupted

**Comment by randomhash** [comment]
good job with syntax checks, tnx

File: .travis.yml (hunk 2 above)
Comment (Enase) [reply .travis.yml:25]: yes, CI will fail in current configuration. so I don't see any things to do here.

File: package.json (hunk 3 above)
Comment (Enase) [reply package.json]: dependencies updated

File: .travis.yml (hunk 2 above)
Comment (Enase) [reply .travis.yml:25]: @randomhash could you explain why it won't fail if you see any case..

**Comment by Enase** [comment]
@randomhash Thanks for review. Please let me know if you see any other suggestions.

File: .travis.yml (hunk 2 above)
Comment (randomhash) [reply .travis.yml:25]: because I saw that lint command will invoke fix command, that should not produce errors ;)

**Review by randomhash — APPROVED** [review]

**Pull request merged**